openssl/src/crypto/pkcs7/pk7_smime.rs

```diff
@@ -0,0 +1,269 @@
+use x509::X509;
+use x509::store::X509Store;
+use ffi;
+use bio::{MemBio, MemBioSlice};
+use error::ErrorStack;
+use stack::Stack;
+use foreign_types::ForeignType;
+use symm::Cipher;
+use pkey::PKey;
+use libc::c_int;
+use std::ptr::null_mut;
+
+pub struct PKCS7(*mut ffi::pkcs7_st);
+
+bitflags! {
+    pub struct PKCS7Flags: c_int {
+        const PKCS7_TEXT = ffi::PKCS7_TEXT;
+        const PKCS7_NOCERTS = ffi::PKCS7_NOCERTS;
+        const PKCS7_NOSIGS = ffi::PKCS7_NOSIGS;
+        const PKCS7_NOCHAIN = ffi::PKCS7_NOCHAIN;
+        const PKCS7_NOINTERN = ffi::PKCS7_NOINTERN;
+        const PKCS7_NOVERIFY = ffi::PKCS7_NOVERIFY;
+        const PKCS7_DETACHED = ffi::PKCS7_DETACHED;
+        const PKCS7_BINARY = ffi::PKCS7_BINARY;
+        const PKCS7_NOATTR = ffi::PKCS7_NOATTR;
+        const PKCS7_NOSMIMECAP = ffi::PKCS7_NOSMIMECAP;
+        const PKCS7_NOOLDMIMETYPE = ffi::PKCS7_NOOLDMIMETYPE;
+        const PKCS7_CRLFEOL = ffi::PKCS7_CRLFEOL;
+        const PKCS7_STREAM = ffi::PKCS7_STREAM;
+        const PKCS7_NOCRL = ffi::PKCS7_NOCRL;
+        const PKCS7_PARTIAL = ffi::PKCS7_PARTIAL;
+        const PKCS7_REUSE_DIGEST = ffi::PKCS7_REUSE_DIGEST;
+        #[cfg(not(any(ossl101, ossl102, libressl)))]
+        const PKCS7_NO_DUAL_CONTENT = ffi::PKCS7_NO_DUAL_CONTENT;
+    }
+}
+
+impl PKCS7 {
+    pub fn smime_write(&self, input: &[u8], flags: PKCS7Flags) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let input_bio = MemBioSlice::new(input)?;
+
+            let output = MemBio::new()?;
+
+            if ffi::SMIME_write_PKCS7(output.as_ptr(), self.0, input_bio.as_ptr(), flags.bits) == 1 {
+                Ok(output.get_buf().to_owned())
+            } else {
+                Err(ErrorStack::get())
+            }
+        }
+    }
+
+    pub fn smime_read(input: &[u8], bcount: &mut Vec<u8>) -> Result<PKCS7, ErrorStack> {
+        let input_bio = MemBioSlice::new(input)?;
+
+        let bcount_bio = MemBio::new()?;
+
+        let pkcs7 = unsafe { ffi::SMIME_read_PKCS7(input_bio.as_ptr(), &mut bcount_bio.as_ptr()) };
+
+        bcount.clear();
+        bcount.append(&mut bcount_bio.get_buf().to_vec());
+
+        if pkcs7.is_null() {
+            Err(ErrorStack::get())
+        } else {
+            Ok(PKCS7(pkcs7))
+        }
+    }
+
+    pub fn decrypt(&self, pkey: &PKey, cert: &X509) -> Result<Vec<u8>, ErrorStack> {
```

Comment:
All you have to do is change `&PKey` to `&PKeyRef` and continue using `.as_ptr()`, same for `X509`.